README.md
# nietzsche.js

![gh-actions](https://github.com/rpidanny/nietzsche.js/workflows/Release/badge.svg) [![codecov](https://codecov.io/gh/rpidanny/nietzsche.js/branch/main/graph/badge.svg?token=UCCA6D6JCB)](https://codecov.io/gh/rpidanny/nietzsche.js)

A JavaScript package to download quotes from Goodreads.

## Install

```sh
$ npm install --save @rpidanny/nietzsche.js
```

## Usage

### Basic example

```ts
import { getAllQuotesByTag } from '@rpidanny/nietzsche.js'

const quoteResponse = await getAllQuotesByTag('psychology')
console.log(quoteResponse)
```

### API Documentation

#### getQuotesByTag(options?)

Type: `object`

| key    | default | type     | description                                                   |
| ------ | ------- | -------- | ------------------------------------------------------------- |
| `tag`  | `None`  | `string` | The name of the tag to search. e.g. `psychology`, `economics` |
| `page` | `1`     | `number` | The page number to get the quotes from.                       |

Response Type: [GetQuotesResponse](#GetQuotesResponse)

#### getAllQuotesByTag(options?)

Type: `object`

| key           | default | type     | description                                                   |
| ------------- | ------- | -------- | ------------------------------------------------------------- |
| `tag`         | `None`  | `string` | The name of the tag to search. e.g. `psychology`, `economics` |
| `concurrency` | `10`    | `number` | The number of parallel requests to goodreads                  |
| `maxPages`    | `100`   | `number` | The maximum number of pages to get quotes from                |

Response Type: `Array<Quotes>`

#### getQuotesByPath(options?)

Type: `object`

| key    | default | type     | description                                                                         |
| ------ | ------- | -------- | ----------------------------------------------------------------------------------- |
| `path` | `None`  | `string` | The relative url path to extract quotes from. e.g. `quotes`, `quotes/tag/economics` |
| `page` | `1`     | `number` | The page number to get the quotes from.                                             |

Response Type: [GetQuotesResponse](#GetQuotesResponse)

#### getAllQuotesByPath(options?)

Type: `object`

| key           | default | type     | description                                                                         |
| ------------- | ------- | -------- | ----------------------------------------------------------------------------------- |
| `path`        | `None`  | `string` | The relative url path to extract quotes from. e.g. `quotes`, `quotes/tag/economics` |
| `concurrency` | `10`    | `number` | The number of parallel requests to goodreads                                        |
| `maxPages`    | `100`   | `number` | The maximum number of pages to get quotes from                                      |

Response Type: `Array<Quotes>`

### Types

#### Quote

```ts
{
  text: string
  author: string
  likes: number
  tags: Array<string>
}
```

#### GetQuotesResponse

```ts
{
  quotes: Array<Quote>
  pageNumber: number
  totalPages: number
  count: number
}
```


CHANGELOG.md
## [1.1.1](https://github.com/rpidanny/nietzsche.js/compare/v1.1.0...v1.1.1) (2021-05-12)


### Bug Fixes

* jest missing types ([#101](https://github.com/rpidanny/nietzsche.js/issues/101)) ([e22ba1a](https://github.com/rpidanny/nietzsche.js/commit/e22ba1ae4a38201dbf1aab387dc44e15763e8253))

# [1.1.0](https://github.com/rpidanny/nietzsche.js/compare/v1.0.0...v1.1.0) (2020-11-07)


### Features

* add get quote by path ([#6](https://github.com/rpidanny/nietzsche.js/issues/6)) ([a0b97f5](https://github.com/rpidanny/nietzsche.js/commit/a0b97f597cfbb16e61e837c8cfcc0b6515d5baaa))

# 1.0.0 (2020-11-04)


### Features

* add download quotes functions ([#1](https://github.com/rpidanny/nietzsche.js/issues/1)) ([780e074](https://github.com/rpidanny/nietzsche.js/commit/780e074dd51b8a211467ee8f9a292a5163032572))

## [1.1.1](https://github.com/rpidanny/typescript-package-template/compare/v1.1.0...v1.1.1) (2020-11-02)


### Bug Fixes

* gh actions badge ([#11](https://github.com/rpidanny/typescript-package-template/issues/11)) ([7524560](https://github.com/rpidanny/typescript-package-template/commit/752456043d0f729fb5481e22334c8c42242a6146))

# [1.1.0](https://github.com/rpidanny/typescript-package-template/compare/v1.0.2...v1.1.0) (2020-10-31)


### Features

* add codecov integration ([#9](https://github.com/rpidanny/typescript-package-template/issues/9)) ([4616955](https://github.com/rpidanny/typescript-package-template/commit/46169557025868e80335f63d94ae5bd43c77641d))

## [1.0.2](https://github.com/rpidanny/typescript-package-template/compare/v1.0.1...v1.0.2) (2020-10-30)


### Bug Fixes

* github-script version ([#8](https://github.com/rpidanny/typescript-package-template/issues/8)) ([83abac7](https://github.com/rpidanny/typescript-package-template/commit/83abac7d2feacc33fc03a5d0bab9e02a885a6d57))

## [1.0.1](https://github.com/rpidanny/typescript-package-template/compare/v1.0.0...v1.0.1) (2020-10-28)


### Bug Fixes

* semantic release issues ([#5](https://github.com/rpidanny/typescript-package-template/issues/5)) ([cda63e5](https://github.com/rpidanny/typescript-package-template/commit/cda63e5dcfd3abb593485bb3e5e800926d6e3219))

# 1.0.0 (2020-10-28)


### Bug Fixes

* gh action ([c621821](https://github.com/rpidanny/typescript-package-template/commit/c621821e658090b94707c8e42fed30924e4c5525))
* github default branch ([#2](https://github.com/rpidanny/typescript-package-template/issues/2)) ([d026266](https://github.com/rpidanny/typescript-package-template/commit/d0262661bd3f75aa7f7bbf435f8591bd77e38a5e))
* github default branch ([#4](https://github.com/rpidanny/typescript-package-template/issues/4)) ([358aea4](https://github.com/rpidanny/typescript-package-template/commit/358aea43738d245d9cb055f10e5d843068c7a17b))


test/unit/parser.spec.ts
import { parseAndExtractQuotes } from '../../src/utils/parser'

describe('Parser', () => {
  const rawHtml =
    '<article>\n<div class=\'quoteContainer quoteContainer--noSeparator\'>\n<div class=\'quoteIconWrapper\'>\n<a aria-label="Orrin Woodward" href="/author/show/249881.Orrin_Woodward"><div class=\'userIcon\' style=\'background-image: url(https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/authors/1359585605i/249881._UY30_.jpg);\'></div>\n</a></div>\n<div class=\'quoteContentContainer\'>\n<div class=\'quote\'>\n<blockquote class=\'quoteBody\'>A limited state with free economic systems is the soil where the liberty tree blossoms.</blockquote>\n<div class=\'quoteDetails\'>\n<span class=\'quoteAuthor\'>\nOrrin Woodward\n</span>\n<div class=\'quoteTags\'>\nTags:\n<a title="Economics quotes" href="/quotes/tag/economics">economics</a>,\n<a title="Freedom quotes" href="/quotes/tag/freedom">freedom</a>,\n<a title="Libertarianism quotes" href="/quotes/tag/libertarianism">libertarianism</a>,\n<a title="Liberty quotes" href="/quotes/tag/liberty">liberty</a>\n</div>\n<div class=\'quoteLikes\'>\n<ul class=\'quoteFooter\'>\n<li class=\'footerItem\'>\n<a id="like_action_quote_567718" rel="nofollow" data-like-count-id="like_count_quote_567718" class="jsLike" data-remote="true" data-method="post" href="/quotes/add/567718-a-limited-state-with-free-economic-systems-is-the-soil?format=json&amp;lc=3">Like</a>\n</li>\n<li class=\'footerItem ratingsCount\'>\n<a id="like_id_quote_567718" href="/quotes/567718-a-limited-state-with-free-economic-systems-is-the-soil"><span class=\'likesIndicator\'>\nlikes:\n</span>\n<span class=\'likesCount\'>3</span>\n</a>\n</li>\n</ul>\n\n</div>\n</div>\n</div>\n\n</div>\n</div>\n\n</article>\n'

  const parsedQuotes = [
    {
      text:
        'A limited state with free economic systems is the soil where the liberty tree blossoms.',
      author: 'Orrin Woodward',
      likes: 3,
      tags: ['economics', 'freedom', 'libertarianism', 'liberty'],
    },
  ]

  it('should parse raw html and extract quotes', () => {
    const quotes = parseAndExtractQuotes(rawHtml)
    expect(quotes).toEqual(parsedQuotes)
  })
})


test/unit/goodreads.spec.ts
import nock from 'nock'
import { performance } from 'perf_hooks'

import { Goodreads } from '../../src/Goodreads'

describe('Goodreads', () => {
  const rawResponse = {
    ok: true,
    content_html:
      '<article>\n<div class=\'quoteContainer quoteContainer--noSeparator\'>\n<div class=\'quoteIconWrapper\'>\n<a aria-label="Orrin Woodward" href="/author/show/249881.Orrin_Woodward"><div class=\'userIcon\' style=\'background-image: url(https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/authors/1359585605i/249881._UY30_.jpg);\'></div>\n</a></div>\n<div class=\'quoteContentContainer\'>\n<div class=\'quote\'>\n<blockquote class=\'quoteBody\'>A limited state with free economic systems is the soil where the liberty tree blossoms.</blockquote>\n<div class=\'quoteDetails\'>\n<span class=\'quoteAuthor\'>\nOrrin Woodward\n</span>\n<div class=\'quoteTags\'>\nTags:\n<a title="Economics quotes" href="/quotes/tag/economics">economics</a>,\n<a title="Freedom quotes" href="/quotes/tag/freedom">freedom</a>,\n<a title="Libertarianism quotes" href="/quotes/tag/libertarianism">libertarianism</a>,\n<a title="Liberty quotes" href="/quotes/tag/liberty">liberty</a>\n</div>\n<div class=\'quoteLikes\'>\n<ul class=\'quoteFooter\'>\n<li class=\'footerItem\'>\n<a id="like_action_quote_567718" rel="nofollow" data-like-count-id="like_count_quote_567718" class="jsLike" data-remote="true" data-method="post" href="/quotes/add/567718-a-limited-state-with-free-economic-systems-is-the-soil?format=json&amp;lc=3">Like</a>\n</li>\n<li class=\'footerItem ratingsCount\'>\n<a id="like_id_quote_567718" href="/quotes/567718-a-limited-state-with-free-economic-systems-is-the-soil"><span class=\'likesIndicator\'>\nlikes:\n</span>\n<span class=\'likesCount\'>3</span>\n</a>\n</li>\n</ul>\n\n</div>\n</div>\n</div>\n\n</div>\n</div>\n\n</article>\n',
    page: 1,
    per_page: 1,
    num_results: 1,
    total_pages: 5,
  }
  const parsedQuotes = [
    {
      text:
        'A limited state with free economic systems is the soil where the liberty tree blossoms.',
      author: 'Orrin Woodward',
      likes: 3,
      tags: ['economics', 'freedom', 'libertarianism', 'liberty'],
    },
  ]

  let gr: Goodreads

  const baseUrl = 'http://example.com'

  beforeEach(async () => {
    gr = new Goodreads(baseUrl)

    nock.abortPendingRequests()
    nock.cleanAll()
    nock.disableNetConnect()
  })

  afterEach(() => {
    const pending = nock.pendingMocks()

    if (pending.length > 0) {
      console.log('Pending Nocks: ', pending)
      throw new Error(`${pending.length} mocks are pending!`)
    }

    nock.enableNetConnect()
  })

  describe('Parser', () => {
    it('should parse raw html and extract quotes', () => {
      const quotes = gr.parseAndExtractQuotes(rawResponse.content_html)
      expect(quotes).toEqual(parsedQuotes)
    })

    it('should filter out quotes with empty text', () => {
      const htmlContent =
        '<article>\n<div class=\'quoteContainer quoteContainer--noSeparator\'>\n<div class=\'quoteIconWrapper\'>\n<a aria-label="Orrin Woodward" href="/author/show/249881.Orrin_Woodward"><div class=\'userIcon\' style=\'background-image: url(https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/authors/1359585605i/249881._UY30_.jpg);\'></div>\n</a></div>\n<div class=\'quoteContentContainer\'>\n<div class=\'quote\'>\n<blockquote class=\'quoteBody\'>A limited state with free economic systems is the soil where the liberty tree blossoms.</blockquote>\n<div class=\'quoteDetails\'>\n<span class=\'quoteAuthor\'></span>\n<div class=\'quoteTags\'>\nTags:\n<a title="Economics quotes" href="/quotes/tag/economics">economics</a>,\n<a title="Freedom quotes" href="/quotes/tag/freedom">freedom</a>,\n<a title="Libertarianism quotes" href="/quotes/tag/libertarianism">libertarianism</a>,\n<a title="Liberty quotes" href="/quotes/tag/liberty">liberty</a>\n</div>\n<div class=\'quoteLikes\'>\n<ul class=\'quoteFooter\'>\n<li class=\'footerItem\'>\n<a id="like_action_quote_567718" rel="nofollow" data-like-count-id="like_count_quote_567718" class="jsLike" data-remote="true" data-method="post" href="/quotes/add/567718-a-limited-state-with-free-economic-systems-is-the-soil?format=json&amp;lc=3">Like</a>\n</li>\n<li class=\'footerItem ratingsCount\'>\n<a id="like_id_quote_567718" href="/quotes/567718-a-limited-state-with-free-economic-systems-is-the-soil"><span class=\'likesIndicator\'>\nlikes:\n</span>\n<span class=\'likesCount\'>3</span>\n</a>\n</li>\n</ul>\n\n</div>\n</div>\n</div>\n\n</div>\n</div>\n\n</article>\n'
      const quotes = gr.parseAndExtractQuotes(htmlContent)
      expect(quotes.length).toEqual(0)
    })
  })

  describe('Quotes by Tag', () => {
    it('should return parsed quotes from a single page', async () => {
      const expectedResponse = {
        quotes: parsedQuotes,
        pageNumber: 1,
        totalPages: 5,
        count: 1,
      }
      nock(baseUrl).get(/.*/).reply(200, rawResponse)
      const response = await gr.getQuotesByTag('economics')
      expect(response).toEqual(expectedResponse)
    })

    it('should return parsed quotes from all pages', async () => {
      const expectedResponse = new Array(5).fill(parsedQuotes[0])
      nock(baseUrl).get(/.*/).times(5).reply(200, rawResponse)
      const response = await gr.getAllQuotesByTag('economics')
      expect(response).toEqual(expectedResponse)
    })

    it('should return parsed quotes from all pages until max pages', async () => {
      const expectedResponse = new Array(2).fill(parsedQuotes[0])
      nock(baseUrl).get(/.*/).times(2).reply(200, rawResponse)
      const response = await gr.getAllQuotesByTag('economics', 10, 2)
      expect(response).toEqual(expectedResponse)
    })

    describe('Request', () => {
      it('should return page with quotes', async () => {
        nock(baseUrl).get(/.*/).reply(200, rawResponse)
        const page = await gr.getPageByTag('economics')
        expect(page).toEqual(rawResponse)
      })

      it('should throw error when failed to get page', async () => {
        nock(baseUrl).get(/.*/).reply(400)
        expect.assertions(2)
        try {
          await gr.getPageByTag('economics')
        } catch (err) {
          expect(err).toBeInstanceOf(Error)
          expect(err).toHaveProperty('message', 'Failed to get page by tag')
        }
      })
    })

    describe('concurrency', () => {
      const delayMillis = 500
      const prescaler = 100

      it('should return parsed quotes from all pages in sequence', async () => {
        const concurrency = 1

        const start = performance.now()

        const expectedResponse = new Array(5).fill(parsedQuotes[0])

        nock(baseUrl).get(/.*/).times(5).delay(delayMillis).reply(200, rawResponse)

        const response = await gr.getAllQuotesByTag('economics', concurrency)

        const end = performance.now()
        const executionTime = end - start

        expect(Math.floor(executionTime / prescaler)).toEqual((5 * delayMillis) / prescaler)
        expect(response).toEqual(expectedResponse)
      })

      it('should return parsed quotes from all pages with concurrent requests', async () => {
        const concurrency = 5

        const start = performance.now()

        const expectedResponse = new Array(5).fill(parsedQuotes[0])

        nock(baseUrl).get(/.*/).times(5).delay(delayMillis).reply(200, rawResponse)

        const response = await gr.getAllQuotesByTag('economics', concurrency)

        const end = performance.now()
        const executionTime = end - start
        // 2 * delayMillis since we need to get the first page before running in parallel
        expect(Math.floor(executionTime / prescaler)).toEqual((2 * delayMillis) / prescaler)
        expect(response).toEqual(expectedResponse)
      })
    })
  })

  describe('Quotes by Path', () => {
    it('should return parsed quotes from a single page', async () => {
      const expectedResponse = {
        quotes: parsedQuotes,
        pageNumber: 1,
        totalPages: 5,
        count: 1,
      }
      nock(baseUrl).get(/.*/).reply(200, rawResponse)
      const response = await gr.getQuotesByPath('quotes/tag/economics')
      expect(response).toEqual(expectedResponse)
    })

    it('should return parsed quotes from all pages', async () => {
      const expectedResponse = new Array(5).fill(parsedQuotes[0])
      nock(baseUrl).get(/.*/).times(5).reply(200, rawResponse)
      const response = await gr.getAllQuotesByPath('quotes/tag/economics')
      expect(response).toEqual(expectedResponse)
    })

    it('should fail when path starts with "/"', async () => {
      expect.assertions(2)
      try {
        await gr.getAllQuotesByPath('/quotes/tag/economics')
      } catch (err) {
        expect(err).toBeInstanceOf(Error)
        expect(err).toHaveProperty(
          'message',
          '`input` must not start with a slash when using `prefixUrl`',
        )
      }
    })

    it('should return parsed quotes from all pages until max pages', async () => {
      const expectedResponse = new Array(2).fill(parsedQuotes[0])
      nock(baseUrl).get(/.*/).times(2).reply(200, rawResponse)
      const response = await gr.getAllQuotesByPath('quotes/tag/economics', 10, 2)
      expect(response).toEqual(expectedResponse)
    })

    describe('Request', () => {
      it('should return page with quotes', async () => {
        nock(baseUrl).get(/.*/).reply(200, rawResponse)
        const page = await gr.getPageByPath('quotes/tag/economics')
        expect(page).toEqual(rawResponse)
      })

      it('should throw error when failed to get page', async () => {
        nock(baseUrl).get(/.*/).reply(400)
        expect.assertions(2)
        try {
          await gr.getPageByPath('quotes/tag/economics')
        } catch (err) {
          expect(err).toBeInstanceOf(Error)
          expect(err).toHaveProperty('message', 'Failed to get page by path')
        }
      })
    })

    describe('concurrency', () => {
      const delayMillis = 500
      const prescaler = 100

      it('should return parsed quotes from all pages in sequence', async () => {
        const concurrency = 1

        const start = performance.now()

        const expectedResponse = new Array(5).fill(parsedQuotes[0])

        nock(baseUrl).get(/.*/).times(5).delay(delayMillis).reply(200, rawResponse)

        const response = await gr.getAllQuotesByPath('quotes/tag/economics', concurrency)

        const end = performance.now()
        const executionTime = end - start

        expect(Math.floor(executionTime / prescaler)).toEqual((5 * delayMillis) / prescaler)
        expect(response).toEqual(expectedResponse)
      })

      it('should return parsed quotes from all pages with concurrent requests', async () => {
        const concurrency = 5

        const start = performance.now()

        const expectedResponse = new Array(5).fill(parsedQuotes[0])

        nock(baseUrl).get(/.*/).times(5).delay(delayMillis).reply(200, rawResponse)

        const response = await gr.getAllQuotesByPath('quotes/tag/economics', concurrency)

        const end = performance.now()
        const executionTime = end - start
        // 2 * delayMillis since we need to get the first page before running in parallel
        expect(Math.floor(executionTime / prescaler)).toEqual((2 * delayMillis) / prescaler)
        expect(response).toEqual(expectedResponse)
      })
    })
  })
})


test/unit/nietzsche.spec.ts
import nock from 'nock'

import { getQuotesByTag, getQuotesByPath, getAllQuotesByTag, getAllQuotesByPath } from '../../src'
import config from '../../src/config'

describe('Nietzsche', () => {
  const baseUrl = config.goodreads.baseUrl
  const rawResponse = {
    ok: true,
    content_html:
      '<article>\n<div class=\'quoteContainer quoteContainer--noSeparator\'>\n<div class=\'quoteIconWrapper\'>\n<a aria-label="Orrin Woodward" href="/author/show/249881.Orrin_Woodward"><div class=\'userIcon\' style=\'background-image: url(https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/authors/1359585605i/249881._UY30_.jpg);\'></div>\n</a></div>\n<div class=\'quoteContentContainer\'>\n<div class=\'quote\'>\n<blockquote class=\'quoteBody\'>A limited state with free economic systems is the soil where the liberty tree blossoms.</blockquote>\n<div class=\'quoteDetails\'>\n<span class=\'quoteAuthor\'>\nOrrin Woodward\n</span>\n<div class=\'quoteTags\'>\nTags:\n<a title="Economics quotes" href="/quotes/tag/economics">economics</a>,\n<a title="Freedom quotes" href="/quotes/tag/freedom">freedom</a>,\n<a title="Libertarianism quotes" href="/quotes/tag/libertarianism">libertarianism</a>,\n<a title="Liberty quotes" href="/quotes/tag/liberty">liberty</a>\n</div>\n<div class=\'quoteLikes\'>\n<ul class=\'quoteFooter\'>\n<li class=\'footerItem\'>\n<a id="like_action_quote_567718" rel="nofollow" data-like-count-id="like_count_quote_567718" class="jsLike" data-remote="true" data-method="post" href="/quotes/add/567718-a-limited-state-with-free-economic-systems-is-the-soil?format=json&amp;lc=3">Like</a>\n</li>\n<li class=\'footerItem ratingsCount\'>\n<a id="like_id_quote_567718" href="/quotes/567718-a-limited-state-with-free-economic-systems-is-the-soil"><span class=\'likesIndicator\'>\nlikes:\n</span>\n<span class=\'likesCount\'>3</span>\n</a>\n</li>\n</ul>\n\n</div>\n</div>\n</div>\n\n</div>\n</div>\n\n</article>\n',
    page: 1,
    per_page: 1,
    num_results: 1,
    total_pages: 5,
  }
  const parsedQuotes = [
    {
      text:
        'A limited state with free economic systems is the soil where the liberty tree blossoms.',
      author: 'Orrin Woodward',
      likes: 3,
      tags: ['economics', 'freedom', 'libertarianism', 'liberty'],
    },
  ]

  describe('Quotes by Tag', () => {
    it('should return parsed quotes from a single page', async () => {
      const expectedResponse = {
        quotes: parsedQuotes,
        pageNumber: 1,
        totalPages: 5,
        count: 1,
      }
      nock(baseUrl).get(/.*/).reply(200, rawResponse)
      const response = await getQuotesByTag({ tag: 'economics' })
      expect(response).toEqual(expectedResponse)
    })

    it('should return parsed quotes from all pages', async () => {
      const expectedResponse = new Array(5).fill(parsedQuotes[0])
      nock(baseUrl).get(/.*/).times(5).reply(200, rawResponse)
      const response = await getAllQuotesByTag({ tag: 'economics' })
      expect(response).toEqual(expectedResponse)
    })
  })

  describe('Quotes by Page', () => {
    it('should return parsed quotes from a single page', async () => {
      const expectedResponse = {
        quotes: parsedQuotes,
        pageNumber: 1,
        totalPages: 5,
        count: 1,
      }
      nock(baseUrl).get(/.*/).reply(200, rawResponse)
      const response = await getQuotesByPath({ path: 'quotes/tag/economics' })
      expect(response).toEqual(expectedResponse)
    })

    it('should return parsed quotes from all pages', async () => {
      const expectedResponse = new Array(5).fill(parsedQuotes[0])
      nock(baseUrl).get(/.*/).times(5).reply(200, rawResponse)
      const response = await getAllQuotesByPath({ path: 'quotes/tag/economics' })
      expect(response).toEqual(expectedResponse)
    })
  })
})


src/index.ts
import { Goodreads } from './Goodreads'
import config from './config'
import * as types from './types'

const gr = new Goodreads(config.goodreads.baseUrl)

export const getQuotesByTag = (
  options: types.GetQuotesByTagRequest,
): Promise<types.GetQuotesResponse> => gr.getQuotesByTag(options.tag, options.page)

export const getAllQuotesByTag = (
  options: types.GetAllQuotesByTagRequest,
): Promise<Array<types.Quote>> =>
  gr.getAllQuotesByTag(options.tag, options.concurrency, options.maxPages)

export const getQuotesByPath = (
  options: types.GetQuotesByPathRequest,
): Promise<types.GetQuotesResponse> => gr.getQuotesByPath(options.path, options.page)

export const getAllQuotesByPath = (
  options: types.GetAllQuotesByPathRequest,
): Promise<Array<types.Quote>> =>
  gr.getAllQuotesByPath(options.path, options.concurrency, options.maxPages)


src/Nietzsche.ts
import $ from 'cheerio'
import got, { Got } from 'got'
import HttpAgent, { HttpsAgent } from 'agentkeepalive'

import * as types from './types'

export class Nietzsche {
  baseUrl: string
  gotInstance: Got
  constructor(baseUrl: string) {
    this.baseUrl = baseUrl
    this.gotInstance = got.extend({
      prefixUrl: this.baseUrl,
      throwHttpErrors: false,
      agent: {
        http: new HttpAgent({
          keepAlive: true,
          maxSockets: 50,
          maxFreeSockets: 10,
          timeout: 60000, // active socket keepalive for 60 seconds
          freeSocketTimeout: 30000, // free socket keepalive for 30 seconds
        }),
        https: new HttpsAgent({
          keepAlive: true,
          maxSockets: 50,
          maxFreeSockets: 10,
          timeout: 60000, // active socket keepalive for 60 seconds
          freeSocketTimeout: 30000, // free socket keepalive for 30 seconds
        }),
      },
    })
  }

  async getPageByTag(tag: string, page = 1): Promise<types.GoodreadsResponse> {
    const response = await this.gotInstance(`quotes/tag/${tag}`, {
      searchParams: {
        format: 'json',
        page,
      },
    })

    if (response.statusCode === 200 && response.body) {
      return JSON.parse(response.body)
    }

    throw new Error('Failed to get page by tag')
  }

  parseAndExtractQuotes(html: string): Array<types.Quote> {
    const quotes: Array<types.Quote> = []
    const quoteGroup = $('.quote', html)

    quoteGroup.each((_idx, elem) => {
      const quoteElement = $(elem)
      const text: string = quoteElement
        .find('.quoteBody')
        .text()
        .replace(/^\s+|\s+$/g, '')
      const author: string = quoteElement
        .find('.quoteAuthor')
        .text()
        .replace(/^\s+|\s+$/g, '')
        .replace(/,/g, '')
      const likes: number = parseInt(
        quoteElement
          .find('.likesCount')
          .text()
          .replace(/^\s+|\s+$/g, ''),
      )
      const tags: Array<string> = quoteElement
        .find('.quoteTags')
        .find('a')
        .map((_idx, tag) => {
          const tagSelector = $(tag)
          return tagSelector.text().replace(/^\s+|\s+$/g, '')
        })
        .toArray()
        .map((element) => element.toString())

      if (text.length > 0 && author.length > 0) {
        const quote: types.Quote = {
          text,
          author,
          likes,
          tags,
        }
        quotes.push(quote)
      }
    })
    return quotes
  }

  async getQuotesByTag(tag: string, pageNumber = 1): Promise<types.GetQuotesResponse> {
    const rawPage = await this.getPageByTag(tag, pageNumber)
    const quotes = this.parseAndExtractQuotes(rawPage.content_html)
    return {
      quotes,
      pageNumber: rawPage.page,
      totalPages: rawPage.total_pages,
      count: quotes.length,
    }
  }

  async getAllQuotesByTag(tag: string, maxPages = 100): Promise<Array<types.Quote>> {
    let nextPage = 1
    let total

    let response: Array<types.Quote> = []
    do {
      const { quotes, totalPages } = await this.getQuotesByTag(tag, nextPage)
      nextPage += 1
      total = totalPages
      response = response.concat(quotes)
    } while (nextPage <= total && nextPage <= maxPages)

    return response
  }
}


src/utils/parser.ts
import $ from 'cheerio'

import * as types from '../types'

export const parseAndExtractQuotes = (html: string): Array<Record<string, unknown>> => {
  const quotes: Array<types.Quote> = []
  const quoteGroup = $('.quote', html)

  quoteGroup.each((_idx, elem) => {
    const quoteElement = $(elem)
    const text = quoteElement
      .find('.quoteBody')
      .text()
      .replace(/^\s+|\s+$/g, '')
    const author = quoteElement
      .find('.quoteAuthor')
      .text()
      .replace(/^\s+|\s+$/g, '')
      .replace(/,/g, '')
    const likes = parseInt(
      quoteElement
        .find('.likesCount')
        .text()
        .replace(/^\s+|\s+$/g, ''),
    )
    const tags = quoteElement
      .find('.quoteTags')
      .find('a')
      .map((_idx, tag) => {
        const tagSelector = $(tag)
        return tagSelector.text().replace(/^\s+|\s+$/g, '')
        // return {
        //   link: tagSelector.attr('href'),
        //   text: tagSelector.text().replace(/^\s+|\s+$/g, '')
        // }
      })
      .toArray()

    if (text.length > 0 && author.length > 0) {
      quotes.push({
        text,
        author,
        likes,
        tags,
      })
    }
  })
  return quotes
}


src/Goodreads.ts
import $ from 'cheerio'
import got, { Got } from 'got'
import HttpAgent, { HttpsAgent } from 'agentkeepalive'

import chunk from './utils/chunk'
import * as types from './types'

export class Goodreads {
  baseUrl: string
  gotInstance: Got
  constructor(baseUrl: string) {
    this.baseUrl = baseUrl
    this.gotInstance = got.extend({
      prefixUrl: this.baseUrl,
      throwHttpErrors: false,
      agent: {
        http: new HttpAgent({
          keepAlive: true,
          maxSockets: 50,
          maxFreeSockets: 10,
          timeout: 60000, // active socket keepalive for 60 seconds
          freeSocketTimeout: 30000, // free socket keepalive for 30 seconds
        }),
        https: new HttpsAgent({
          keepAlive: true,
          maxSockets: 50,
          maxFreeSockets: 10,
          timeout: 60000, // active socket keepalive for 60 seconds
          freeSocketTimeout: 30000, // free socket keepalive for 30 seconds
        }),
      },
    })
  }

  async getPageByTag(tag: string, page = 1): Promise<types.GoodreadsResponse> {
    const response = await this.gotInstance(`quotes/tag/${tag}`, {
      searchParams: {
        format: 'json',
        page,
      },
    })

    if (response.statusCode === 200 && response.body) {
      return JSON.parse(response.body)
    }

    throw new Error('Failed to get page by tag')
  }

  async getPageByPath(path: string, page = 1): Promise<types.GoodreadsResponse> {
    const response = await this.gotInstance(path, {
      searchParams: {
        format: 'json',
        page,
      },
    })

    if (response.statusCode === 200 && response.body) {
      return JSON.parse(response.body)
    }

    throw new Error('Failed to get page by path')
  }

  parseAndExtractQuotes(html: string): Array<types.Quote> {
    const quotes: Array<types.Quote> = []
    const quoteGroup = $('.quote', html)

    quoteGroup.each((_idx, elem) => {
      const quoteElement = $(elem)
      const text: string = quoteElement
        .find('.quoteBody')
        .text()
        .replace(/^\s+|\s+$/g, '')
      const author: string = quoteElement
        .find('.quoteAuthor')
        .text()
        .replace(/^\s+|\s+$/g, '')
        .replace(/,/g, '')
      const likes: number = parseInt(
        quoteElement
          .find('.likesCount')
          .text()
          .replace(/^\s+|\s+$/g, ''),
      )
      const tags: Array<string> = quoteElement
        .find('.quoteTags')
        .find('a')
        .map((_idx, tag) => {
          const tagSelector = $(tag)
          return tagSelector.text().replace(/^\s+|\s+$/g, '')
        })
        .toArray()
        .map((element) => element.toString())

      if (text.length > 0 && author.length > 0) {
        const quote: types.Quote = {
          text,
          author,
          likes,
          tags,
        }
        quotes.push(quote)
      }
    })
    return quotes
  }

  async getQuotesByTag(tag: string, pageNumber = 1): Promise<types.GetQuotesResponse> {
    const rawPage = await this.getPageByTag(tag, pageNumber)
    const quotes = this.parseAndExtractQuotes(rawPage.content_html)
    return {
      quotes,
      pageNumber: rawPage.page,
      totalPages: rawPage.total_pages,
      count: quotes.length,
    }
  }

  async getQuotesByPath(path: string, pageNumber = 1): Promise<types.GetQuotesResponse> {
    const rawPage = await this.getPageByPath(path, pageNumber)
    const quotes = this.parseAndExtractQuotes(rawPage.content_html)
    return {
      quotes,
      pageNumber: rawPage.page,
      totalPages: rawPage.total_pages,
      count: quotes.length,
    }
  }

  private async getAllQuotes(
    method: (selector: string, pageNumber: number) => Promise<types.GetQuotesResponse>,
    selector: string,
    concurrency = 10,
    maxPages = 100,
  ): Promise<Array<types.Quote>> {
    let response: Array<types.Quote> = []

    // download first page to get total page info
    const { quotes, totalPages } = await method(selector, 1)

    response = response.concat(quotes)

    // Create an array of page numbers
    const pageNumbers: Array<number> = Array.from(
      { length: totalPages < maxPages ? totalPages : maxPages },
      (_, i) => i + 1,
    )

    // remove the first page as it has already been downloaded
    pageNumbers.splice(0, 1)

    // create batches to run in parallel
    const batches = chunk(pageNumbers, concurrency) as Array<Array<number>>

    for (let i = 0; i < batches.length; i++) {
      const batch = batches[i]
      const responses = await Promise.all(
        batch.map((pageNumber: number) => method(selector, pageNumber)),
      )

      const batchQuotes = responses.reduce((acc, item) => {
        acc = acc.concat(item.quotes)
        return acc
      }, [] as Array<types.Quote>)
      response = response.concat(batchQuotes)
    }

    return response
  }

  async getAllQuotesByTag(
    tag: string,
    concurrency = 10,
    maxPages = 100,
  ): Promise<Array<types.Quote>> {
    const quotes = await this.getAllQuotes(
      this.getQuotesByTag.bind(this),
      tag,
      concurrency,
      maxPages,
    )
    return quotes
  }

  async getAllQuotesByPath(
    path: string,
    concurrency = 10,
    maxPages = 100,
  ): Promise<Array<types.Quote>> {
    const quotes = await this.getAllQuotes(
      this.getQuotesByPath.bind(this),
      path,
      concurrency,
      maxPages,
    )
    return quotes
  }
}
